File: python_backend/getters/get_num_pieces_checking_king.py
from threatArea.get_knight_threats import get_knight_threats
from threatArea.get_pawn_threats import get_pawn_threats
from pathsInfo.top.get_pathdata_dict import get_pathdata_dict
from getters.get_num_path_attackers import get_num_path_attackers


def get_num_pieces_checking_king(sqr, board, color, pd_dict):
    """get the number of pieces checking the king of current turn"""
    knight_threats = get_knight_threats(sqr, board, color)
    pawn_threats = get_pawn_threats(sqr, board, color)
    num_path_attackers = get_num_path_attackers(pd_dict)
    return len(knight_threats) + len(pawn_threats) + num_path_attackers


if __name__ == "__main__":
    # test #1, check_example3, 3k3b/q2r2P1/3P1P2/8/q1QK1QBr/2NP4/3R1R2/b2q2q1
    print('test 1:')
    board = \
        {(1, 1): 'BB1', (1, 2): '#', (1, 3): '#', (1, 4): 'BQ3', (1, 5): '#', (1, 6): '#', (1, 7): 'BQ4', (1, 8): '#',
         (2, 1): '#', (2, 2): '#', (2, 3): '#', (2, 4): '#', (2, 5): '#', (2, 6): '#', (2, 7): '#', (2, 8): '#',
         (3, 1): '#', (3, 2): '#', (3, 3): 'WN1', (3, 4): 'WQ1', (3, 5): '#', (3, 6): '#', (3, 7): '#', (3, 8): '#',
         (4, 1): 'BQ1', (4, 2): 'WR1', (4, 3): 'WP1', (4, 4): 'WK1', (4, 5): '#', (4, 6): 'WP2', (4, 7): 'BR2',
         (4, 8): 'BK1',
         (5, 1): '#', (5, 2): '#', (5, 3): '#', (5, 4): '#', (5, 5): '#', (5, 6): '#', (5, 7): '#', (5, 8): '#',
         (6, 1): '#', (6, 2): 'WR2', (6, 3): '#', (6, 4): 'WQ2', (6, 5): '#', (6, 6): 'WP3', (6, 7): '#', (6, 8): '#',
         (7, 1): 'BQ2', (7, 2): '#', (7, 3): '#', (7, 4): 'WB1', (7, 5): '#', (7, 6): '#', (7, 7): 'WP4', (7, 8): '#',
         (8, 1): '#', (8, 2): '#', (8, 3): '#', (8, 4): 'BR1', (8, 5): '#', (8, 6): '#', (8, 7): '#', (8, 8): 'BB2'}

    pd_dict = get_pathdata_dict((4, 4), board, 'W')
    print(get_num_pieces_checking_king((4, 4), board, 'W', pd_dict))
    print('\n')

    # test 5, super_checkmate_impossible_example, 5rk1/3np1p1/r4K2/8/7b/8/8/q4r2:
    print('test 2:')
    board = \
        {(1, 8): '#', (2, 8): '#', (3, 8): '#', (4, 8): '#', (5, 8): '#', (6, 8): 'BR3', (7, 8): 'BK1', (8, 8): '#',
         (1, 7): '#', (2, 7): '#', (3, 7): '#', (4, 7): 'BN1', (5, 7): 'BP1', (6, 7): '#', (7, 7): 'BP2', (8, 7): '#',
         (1, 6): 'BR2', (2, 6): '#', (3, 6): '#', (4, 6): '#', (5, 6): '#', (6, 6): 'WK1', (7, 6): '#', (8, 6): '#',
         (1, 5): '#', (2, 5): '#', (3, 5): '#', (4, 5): '#', (5, 5): '#', (6, 5): '#', (7, 5): '#', (8, 5): '#',
         (1, 4): '#', (2, 4): '#', (3, 4): '#', (4, 4): '#', (5, 4): '#', (6, 4): '#', (7, 4): '#', (8, 4): 'BB1',
         (1, 3): '#', (2, 3): '#', (3, 3): '#', (4, 3): '#', (5, 3): '#', (6, 3): '#', (7, 3): '#', (8, 3): '#',
         (1, 2): '#', (2, 2): '#', (3, 2): '#', (4, 2): '#', (5, 2): '#', (6, 2): '#', (7, 2): '#', (8, 2): '#',
         (1, 1): 'BQ1', (2, 1): '#', (3, 1): '#', (4, 1): '#', (5, 1): '#', (6, 1): 'BR1', (7, 1): '#', (8, 1): '#'}

    pd_dict = get_pathdata_dict((6, 6), board, 'W')
    print(get_num_pieces_checking_king((6, 6), board, 'W', pd_dict))
    print('\n')

File: node.js/restriction/getNumPiecesCheckingKing.js
const  getOffsetThreats  = require("../threatArea/getOffsetThreats");
const  getPawnThreats  = require("../threatArea/getPawnThreats");
const  getNumPathAttackers  = require("../pathsInfo/top/getNumPathAttackers");



function getNumPiecesCheckingKing(board, sqr, color, pdDict, pieceDefs, idDict) {
    /*get the number of pieces checking the king of current turn**/
    let offsetThreats = getOffsetThreats(board, sqr, color, pieceDefs, idDict)
    let pawnThreats = getPawnThreats(board, sqr, color)
    let numPathAttackers = getNumPathAttackers(pdDict)
    return offsetThreats.length + pawnThreats.length + numPathAttackers
}

module.exports = getNumPiecesCheckingKing;

if (require.main === module) {

    const  getPathdataDict  = require("../pathsInfo/top/getpathDataDict");
    const  getStandardPieceDefs  = require("../testObjects/getStandardPieceDefs");
    const  getStandardIdDict  = require("../testObjects/getStandardIdDict");
    const  sampleBoardDicts  = require("../testObjects/sampleBoardDicts");
    const  printBoard  = require("../printers/printBoard");
    const initEmptyRanges = require("../apiHelpers/initEmptyRanges");
    const getPieceLoc = require(".././helpers/getPieceLoc")
    const {rfToXy} = require("../coordType/crdCnvrt")

    var pieceDefs = initEmptyRanges(getStandardPieceDefs())
    var idDict = getStandardIdDict()

    var pdDict, board, n, kingLoc;

    var game_names = [
    'check_example3',
    'fundemental_defense',
    'king_range_test',
    'knight_threat2',
    'pawn_threat1',
    'pinned_ex2',
    'stalemate_exp1',
    'super_checkmate_impossible_example']

    for (var game_name of game_names) {
        board = sampleBoardDicts[game_name]
        kingLoc = getPieceLoc(board, 'WK1')
        pdDict = getPathdataDict(board, rfToXy(kingLoc), 'W', pieceDefs, idDict)
        n = getNumPiecesCheckingKing(board, rfToXy(kingLoc), 'W', pdDict, pieceDefs, idDict)
        printBoard(board, `\n${game_name}, ${n}`, [kingLoc])
    }
}